common/kvstore/ttl/ttl_wrapper.go

```diff
@@ -0,0 +1,305 @@
+func (store *ttlStore) PutBatchWithTTL(keys [][]byte, values [][]byte, ttl time.Duration) error {
+	expiryTime := time.Now().Add(ttl)
+	return store.PutBatchWithExpiration(keys, values, expiryTime)
+}
+
+// buildExpiryKey creates an expiry key from the given expiry time.
+// The expiry key is composed of the following 3 components appended one after the other:
+// - a one byte "e" prefix
+// - the expiry time in hexadecimal format (8 bytes)
+// - and the base key.
+func buildExpiryKey(
+	baseKey []byte,
+	expiryTime time.Time) []byte {
+
+	expiryKeyLength := 1 /* prefix */ + 8 /* expiry timestamp */ + len(baseKey)
+	expiryKey := make([]byte, expiryKeyLength)
+
+	expiryKey[0] = 'e'
+	expiryUnixNano := expiryTime.UnixNano()
+	for i := 0; i < 8; i++ {
+		expiryKey[1+i] = byte((expiryUnixNano >> (56 - i*8)) & 0xff)
+	}
+
+	copy(expiryKey[9:], baseKey)
+
+	return expiryKey
+}
+
+// parseExpiryKey extracts the expiry time and base key from the given expiry key.
+func parseExpiryKey(expiryKey []byte) (baseKey []byte, expiryTime time.Time) {
+	expiryUnixNano := int64(0)
+	for i := 0; i < 8; i++ {
+		expiryUnixNano |= int64(expiryKey[1+i]) << (56 - i*8)
+	}
+	expiryTime = time.Unix(0, expiryUnixNano)
+
+	baseKey = expiryKey[9:]
+	return
+}
+
+// PutWithExpiration adds a key-value pair to the store that expires at a specified time.
+// Key is eventually deleted after the expiry time.
+func (store *ttlStore) PutWithExpiration(key []byte, value []byte, expiryTime time.Time) error {
+	prefixedKey := append(keyPrefix, key...)
+
+	batchKeys := make([][]byte, 2)
+	batchValues := make([][]byte, 2)
+
+	batchKeys[0] = prefixedKey
+	batchValues[0] = value
+
+	batchKeys[1] = buildExpiryKey(key, expiryTime)
+	batchValues[1] = make([]byte, 0)
+
+	return store.store.WriteBatch(batchKeys, batchValues)
+}
+
+// PutBatchWithExpiration adds multiple key-value pairs to the store that expire at a specified time.
+func (store *ttlStore) PutBatchWithExpiration(keys [][]byte, values [][]byte, expiryTime time.Time) error {
+	if len(keys) != len(values) {
+		return fmt.Errorf("keys and values must have the same length (keys: %d, values: %d)", len(keys), len(values))
+	}
+
+	batchKeys := make([][]byte, len(keys)*2)
+	batchValues := make([][]byte, len(keys)*2)
+
+	for i, key := range keys {
+		prefixedKey := append(keyPrefix, key...)
+
+		batchKeys[i*2] = prefixedKey
+		batchValues[i*2] = values[i]
+
+		batchKeys[i*2+1] = buildExpiryKey(key, expiryTime)
+		batchValues[i*2+1] = make([]byte, 0)
+	}
+
+	return store.store.WriteBatch(batchKeys, batchValues)
+}
+
+// Spawns a background goroutine that periodically checks for expired keys and deletes them.
+func (store *ttlStore) expireKeysInBackground(gcPeriod time.Duration) {
+	ticker := time.NewTicker(gcPeriod)
+	go func() {
+		for {
+			select {
+			case now := <-ticker.C:
+				err := store.expireKeys(now)
+				if err != nil {
```

Comment:
Stopping the GC upon an error may be not desired, it may have to continue as an error is not fatal to GC

Reply:
This will now log an error without shutting things down.